### ISSUE TITLE
"Argument unused during compilation" errors on AArch64 assembly files in 0.12.1

### ISSUE THREAD

**Issue by Torphedo** [ziglang/zig#20416]
### Zig Version

0.12.1

### Steps to Reproduce and Observed Behavior

I started running into this problem after updating from `0.11.0`. After updating my build script to use the new `ResolvedTarget` struct, Zig started throwing errors about unused Clang arguments. In my codebase I also have an `svc.s` containing syscall wrappers, and the same errors were thrown for that file (which leads me to believe it has to do with assembly files).

I also tried adding the C flag `-Wno-error` for `main.c`, and even adding the assembly file as a C source with that flag, but nothing changed. Even when adding the assembly file as a C source (which should just error out when it realizes the file is assembly), it only threw errors about the unused compiler flags.

File `build.zig`:
```zig
const std = @import("std");

pub fn build(b: *std.Build) void {
    const switch_target = b.resolveTargetQuery(.{
        .cpu_arch = .aarch64,
        .cpu_model = .{ .explicit = &std.Target.aarch64.cpu.cortex_a57 },
        .os_tag = .freestanding,
        .abi = .none,
        .ofmt = .elf,
    });

    const exe = b.addExecutable(.{
        .name = "zig-hos",
        .target = switch_target,
        .link_libc = false,
        .optimize = b.standardOptimizeOption(.{}),
    });
    exe.addCSourceFile(.{ .file = b.path("main.c"), .flags = &.{} });

    exe.addAssemblyFile(b.path("nro_entry.s"));
    b.installArtifact(exe);
}
```
File `main.c`:
```c
int main() {
    return 0;
}
```
File `nro_entry.s`:
```asm
.section ".text.crt0","ax"

.global _start
_start:
    b main

    // Put MOD0 offset in the NRO file.
    .word __nx_mod0 - _start

    // MOD0 has to start after the NRO header and at a power of 2 offset
    .balign 0x80

__nx_mod0:
    .ascii "MOD0"
    .word  __dynamic_start__        - __nx_mod0
    .word  __bss_start__            - __nx_mod0
    .word  __bss_end__              - __nx_mod0
    .word  __eh_frame_hdr_start__   - __nx_mod0
    .word  __eh_frame_hdr_end__     - __nx_mod0
    .word  main - __nx_mod0

// Put the rest of the code at the next 0x10 boundary just to make it neater
.balign 0x10

// Other code would go here, removed because it didn't affect the build errors
```

Output:
```
install
+- install zig-hos
   +- zig build-exe zig-hos Debug aarch64-freestanding-none 3 errors
:1:1: error: argument unused during compilation: '-fno-caret-diagnostics'
:1:1: error: argument unused during compilation: '-ffreestanding'
:1:1: error: argument unused during compilation: '--serialize-diagnostics /home/torph/dev/compile/zig-cache/tmp/31ba99a6cbfcf23-nro_entry.o.diag'
error: the following command failed with 3 compilation errors:
/usr/bin/zig build-exe -ofmt=elf /home/torph/dev/compile/main.c /home/torph/dev/compile/nro_entry.s -ODebug -target aarch64-freestanding-none -mcpu cortex_a57 -Mroot --cache-dir /home/torph/dev/compile/zig-cache --global-cache-dir /home/torph/.cache/zig --name zig-hos --listen=- 
Build Summary: 0/3 steps succeeded; 1 failed (disable with --summary none)
install transitive failure
+- install zig-hos transitive failure
   +- zig build-exe zig-hos Debug aarch64-freestanding-none 3 errors
error: the following build command failed with exit code 1:
/home/torph/dev/compile/zig-cache/o/b636d19a996bb823a2f9502849950650/build /usr/bin/zig /home/torph/dev/compile /home/torph/dev/compile/zig-cache /home/torph/.cache/zig --seed 0xba0a69a8 -Z37049a5be179e482
```

### Expected Behavior

I would expect these flags not to be passed if they're not needed, or for it to be treated as a warning. When I searched for other people having this issue, it always said Clang reported it as a warning, not an error.

**Comment by rohlem** [comment]
> I also tried adding the C flag `-Wno-error` for `main.c`, and even adding the assembly file as a C source with that flag, but nothing changed. 

Zig treats any warning coming from clang as an error.
The way I understand it, `-Wno-error` tells clang that warnings aren't errors, however Zig is the one taking an issue with them, even if they're just warnings.
To suppress the clang warning entirely, it seems like [command line option `-Qunused-arguments`](https://releases.llvm.org/18.1.0/tools/clang/docs/ClangCommandLineReference.html#cmdoption-clang-Qunused-arguments) should work.

(However, that's more of a workaround; of course the nicer thing would be, as you wrote, not to pass those flags, whatever they're caused by.
Since they're not mentioned anywhere in your given build script, my guess would be that they may come from a `#pragma` directive or something similar in one of the files that get included, but maybe I'm wrong on that.)

As for adding the flag, on the C file you can add `"-Qunused-arguments"` directly to the passed `.flags` field.
Because the third entry mentions `nro_entry` though, I agree the main issue might be the assembly file.
Currently I don't see a convenient way to add flags to assembly files via the std.Build API though, so you might want to try hardcoding adding the flag in [`std.Build.Step.Compile.make`](https://github.com/ziglang/zig/blob/3e9ab6aa7b2d90c25cb906d425a148abf9da3dcb/lib/std/Build/Step/Compile.zig#L1215-L1219) of the `lib/std` directory in your Zig installation.
As a starting point, if I'm reading the code correctly, [we currently strip all previous `-cflags` from compiled assembly files](https://github.com/ziglang/zig/blob/3e9ab6aa7b2d90c25cb906d425a148abf9da3dcb/lib/std/Build/Step/Compile.zig#L1215-L1219). Instead, you'd probably want `try zig_args.append("--cflags -Qunused-arguments --");` and set `prev_has_cflags = true;` [like it's done for `.c_source_file` below](https://github.com/ziglang/zig/blob/3e9ab6aa7b2d90c25cb906d425a148abf9da3dcb/lib/std/Build/Step/Compile.zig#L1234-L1239).

**Comment by Torphedo** [comment]
> The way I understand it, -Wno-error tells clang that warnings aren't errors, however Zig is the one taking an issue with them, even if they're just warnings.
I see, that explains why the flag wasn't helping.

> you might want to try hardcoding adding the flag in [std.Build.Step.Compile.make](https://github.com/ziglang/zig/blob/3e9ab6aa7b2d90c25cb906d425a148abf9da3dcb/lib/std/Build/Step/Compile.zig#L1215-L1219) of the lib/std directory in your Zig installation.
This is obviously not ideal, but editing `std/Build/Step/Compile.zig` was enough to hack it into at least compiling. I guess that's the nice thing about your toolchain shipping with full source code!

Hopefully this can be fixed properly in the main branch, I can't easily fall back to `0.11.0` because the [libc implementation](https://github.com/kassane/picolibc-zig) I want to use only has a `0.12.0` build script.

**Comment by Torphedo** [comment]
Something strange happened... I wrote a workaround using `std.ChildProcess` to manually run `zig build-obj` with the correct flags, and add the object file with `.addObjectFile()`. But when I removed my edits to the stdlib, everything worked fine with the normal `.addAssemblyFile()`. I uninstalled Zig, cleared my package manager caches, and re-downloaded it, and it still worked fine. Even the separate folder I made with the code above to reproduce the issue compiled fine with the unmodified stdlib.

So... I don't know what happened, but everything is fine now? I guess if it breaks again, I have a workaround at the ready...

**Comment by alexrp** [comment]
Did you wipe `~/.cache/zig`?

**Comment by Torphedo** [comment]
> Did you wipe `~/.cache/zig`?

I didn't notice that folder, but nothing changed when I did just now. I also pushed my changes to git and then fresh installed Zig on my home server. I git cloned there and it built first try with no issues.

**Comment by jacobly0** [comment]
Already fixed by #20068 which is part of the 0.13.0 release, but not the 0.12.1 release.
